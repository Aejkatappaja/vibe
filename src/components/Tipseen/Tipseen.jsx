import { useRef, forwardRef, useMemo } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import useMergeRefs from "../../hooks/useMergeRefs";
import { DialogPositions } from "../../constants/sizes";
import Tooltip from "../Tooltip/Tooltip";
import { DIALOG_ANIMATION_TYPES } from "../../constants/AnimationTypes";
import Button from "../Button/Button";
import { BEMClass } from "../../helpers/bem-helper";
import Icon from "../Icon/Icon";
import CloseSmall from "../Icon/Icons/components/CloseSmall";
import "./Tipseen.scss";
import { TOOLTIP_JUSTIFY_TYPES } from "../Tooltip/TooltipConstants";

const TIPSEEN_BASE_CSS_CLASS = "monday-style-tipseen";
const bemHelper = BEMClass(TIPSEEN_BASE_CSS_CLASS);

const Tipseen = forwardRef(
  (
    {
      className,
      id,
      position,
      animationType,
      hideDelay,
      showDelay,
      title,
      isCloseButtonHidden,
      onClose,
      closeAriaLabel,
      children,
      content,
      justify,
      isCloseButtonOnImage,
      containerSelector
    },
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });
    const overrideCloseAriaLabel = closeAriaLabel || "Close";
    const tipseenContent = useMemo(
      () => (
        <div className={TIPSEEN_BASE_CSS_CLASS}>
          <div className={bemHelper({ element: "header" })}>
            <Button
              className={bemHelper({ element: "close-button" })}
              onClick={onClose}
              size={Button.sizes.SMALL}
              kind={Button.kinds.TERTIARY}
              color={isCloseButtonOnImage ? Button.colors.ON_INVERTED_BACKGROUND : Button.colors.ON_PRIMARY_COLOR}
              ariaLabel={overrideCloseAriaLabel}
            >
              {isCloseButtonHidden ? null : <Icon clickable={false} icon={CloseSmall} iconSize={20} ignoreFocusStyle />}
            </Button>
            {title ? <span className={bemHelper({ element: "title" })}>{title}</span> : null}
          </div>
          <div className={bemHelper({ element: "content" })}>{content}</div>
        </div>
      ),
      [content, isCloseButtonHidden, isCloseButtonOnImage, onClose, overrideCloseAriaLabel, title]
    );

    return (
      <div ref={mergedRef} id={id}>
        <Tooltip
          className={cx(`${TIPSEEN_BASE_CSS_CLASS}-wrapper`, className)}
          shouldShowOnMount
          position={position}
          animationType={animationType}
          hideDelay={hideDelay}
          showDelay={showDelay}
          content={tipseenContent}
          theme={Tooltip.themes.Primary}
          showTrigger={["click"]}
          hideTrigger={["click"]}
          justify={justify}
          containerSelector={containerSelector}
        >
          {children}
        </Tooltip>
      </div>
    );
  }
);

Tipseen.positions = DialogPositions;
Tipseen.animationTypes = DIALOG_ANIMATION_TYPES;
Tipseen.justifyTypes = TOOLTIP_JUSTIFY_TYPES;
Tipseen.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string,
  position: PropTypes.oneOf([
    Tipseen.positions.LEFT,
    Tipseen.positions.LEFT_START,
    Tipseen.positions.LEFT_END,
    Tipseen.positions.RIGHT,
    Tipseen.positions.RIGHT_START,
    Tipseen.positions.RIGHT_END,
    Tipseen.positions.TOP,
    Tipseen.positions.TOP_START,
    Tipseen.positions.TOP_END,
    Tipseen.positions.BOTTOM,
    Tipseen.positions.BOTTOM_START,
    Tipseen.positions.BOTTOM_END
  ]),
  animationType: PropTypes.oneOf([Tipseen.animationTypes.EXPAND, Tipseen.animationTypes.OPACITY_AND_SLIDE]),
  hideDelay: PropTypes.number,
  showDelay: PropTypes.number,
  title: PropTypes.string,
  isCloseButtonHidden: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  containerSelector: PropTypes.string,
  justify: PropTypes.oneOf([Tipseen.justifyTypes.START, Tipseen.justifyTypes.CENTER, Tipseen.justifyTypes.END])
};
Tipseen.defaultProps = {
  className: "",
  id: "",
  position: Tipseen.positions.BOTTOM,
  animationType: Tipseen.animationTypes.EXPAND,
  hideDelay: 0,
  showDelay: 0,
  title: undefined,
  isCloseButtonHidden: false,
  children: null,
  containerSelector: undefined,
  justify: Tipseen.justifyTypes.CENTER
};

export default Tipseen;
