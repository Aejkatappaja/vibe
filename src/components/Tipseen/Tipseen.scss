@import "../../styles/themes.scss";
@import "../../styles/typography.scss";
@import "../../styles/global-css-settings.scss";

@mixin tipseen-title {
  @include font-paragraph-bold;
  color: var(--text-color-on-primary);
}
.monday-style-tipseen {
  &-wrapper {
    min-width: 100px;
    min-height: 40px;
    padding: $spacing-medium;
  }

  &_header {
    display: flex;
    justify-content: flex-start;
  }

  &_close-button {
    position: absolute;
    top: $spacing-xs;
    right: $spacing-xs;
  }

  &_title {
    @include tipseen-title;
  }
}