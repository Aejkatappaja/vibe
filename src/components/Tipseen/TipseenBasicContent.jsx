import cx from "classnames";
import PropTypes from "prop-types";
import TipseenTitle from "./TipseenTitle";
import "./TipseenBasicContent.scss";

const BASE_CSS_CLASS = "monday-style-tipseen-basic-content";

const TipseenBasicContent = ({ title, children, className }) => {
  return (
    <div className={cx(BASE_CSS_CLASS, className)}>
      <TipseenTitle text={title} />
      {children}
    </div>
  );
};

TipseenBasicContent.propTypes = {
  title: PropTypes.string,
  children: PropTypes.element
};

TipseenBasicContent.defaultProps = {
  title: undefined,
  children: PropTypes.element
};

export default TipseenBasicContent;
