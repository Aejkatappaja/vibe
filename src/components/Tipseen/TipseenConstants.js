export const TIPSEEN_ARROW_POSITION = {};

export const TIPSEEN_ANIMATION_TYPE = {};
