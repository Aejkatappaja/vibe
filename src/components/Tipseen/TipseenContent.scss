@import "../../styles/themes.scss";
@import "../../styles/typography.scss";
@import "../../styles/global-css-settings.scss";

.monday-style-tipseen-content {
  display: flex;
  flex-direction: column;
  &_buttons {
    margin-top: $spacing-small;
    display: flex;
    justify-content: flex-end;
    .monday-style-tipseen-content_dismiss {
      margin-right: $spacing-medium;
    }
  }
}