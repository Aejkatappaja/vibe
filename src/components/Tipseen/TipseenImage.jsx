import cx from "classnames";
import "./TipseenImage.scss";

const BASE_CSS_CLASS = "monday-style-tipseen-image";
const TipseenImage = ({ src, alt, className }) => {
  return <img src={src} alt={alt} className={cx(BASE_CSS_CLASS, className)} />;
};

export default TipseenImage;
