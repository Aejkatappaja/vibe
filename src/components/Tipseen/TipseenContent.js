import React from "react";
import cx from "classnames";
import PropTypes from "prop-types";
import { BEMClass } from "../../helpers/bem-helper";
import { NOOP } from "../../utils/function-utils";
import { Button } from "../index";
import { DISMISS_BUTTON_TEXT, SUBMIT_BUTTON_TEXT } from "./TipseenConstants";
import "./TipseenContent.scss";

const BASE_CSS_CLASS = "monday-style-tipseen-content";

const bemHelper = BEMClass(BASE_CSS_CLASS);
const TipseenContent = ({ title, content, isDismissHidden, isSubmitHidden, dismissButtonProps, submitButtonProps }) => {
  const { content: dismissContent, className: dismissClassName, ...otherDismissButtonProps } = dismissButtonProps;
  const { content: submitContent, className: submitClassName, ...otherSubmitButtonProps } = submitButtonProps;
  return (
    <div className={BASE_CSS_CLASS}>
      {title ? <span className={bemHelper({ element: "title" })}>{title}</span> : null}
      {content ? <span className={bemHelper({ element: "content" })}>{content}</span> : null}
      <div className={bemHelper({ element: "buttons" })}>
        {isDismissHidden ? null : (
          <Button
            kind={Button.kinds.TERTIARY}
            color={Button.colors.ON_PRIMARY_COLOR}
            className={cx(bemHelper({ element: "dismiss" }), dismissClassName)}
            size={Button.sizes.SMALL}
            {...otherDismissButtonProps}
          >
            {dismissContent || DISMISS_BUTTON_TEXT}
          </Button>
        )}
        {isSubmitHidden ? null : (
          <Button
            kind={Button.kinds.PRIMARY}
            color={Button.colors.ON_PRIMARY_COLOR}
            size={Button.sizes.SMALL}
            className={cx(bemHelper({ element: "submit" }), submitClassName)}
            {...otherSubmitButtonProps}
          >
            {submitContent || SUBMIT_BUTTON_TEXT}
          </Button>
        )}
      </div>
    </div>
  );
};

TipseenContent.propTypes = {
  title: PropTypes.string,
  content: PropTypes.oneOf([PropTypes.element, PropTypes.string]),
  onSubmit: PropTypes.func,
  onDismiss: PropTypes.func,
  isDismissHidden: PropTypes.bool,
  isSubmitHidden: PropTypes.bool,
  // eslint-disable-next-line react/forbid-prop-types
  dismissButtonProps: PropTypes.object,
  // eslint-disable-next-line react/forbid-prop-types
  submitButtonProps: PropTypes.object
};

TipseenContent.defaultProps = {
  title: undefined,
  content: undefined,
  onSubmit: NOOP,
  onDismiss: NOOP,
  isDismissHidden: true,
  isSubmitHidden: false,
  dismissButtonProps: {},
  submitButtonProps: {}
};

export default TipseenContent;
