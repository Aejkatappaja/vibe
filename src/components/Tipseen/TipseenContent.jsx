import cx from "classnames";
import PropTypes from "prop-types";
import { BEMClass } from "../../helpers/bem-helper";
import { NOOP } from "../../utils/function-utils";
import { Button } from "../index";
import { DISMISS_BUTTON_TEXT, SUBMIT_BUTTON_TEXT } from "./TipseenConstants";
import TipseenBasicContent from "./TipseenBasicContent";
import "./TipseenContent.scss";

const BASE_CSS_CLASS = "monday-style-tipseen-content";

const bemHelper = BEMClass(BASE_CSS_CLASS);
const TipseenContent = ({
  title,
  children,
  isDismissHidden,
  isSubmitHidden,
  dismissButtonProps,
  submitButtonProps
}) => {
  console.log(children);
  const { content: dismissContent, className: dismissClassName, ...otherDismissButtonProps } = dismissButtonProps;
  const { content: submitContent, className: submitClassName, ...otherSubmitButtonProps } = submitButtonProps;
  return (
    <TipseenBasicContent title={title} className={BASE_CSS_CLASS}>
      {children ? <span className={bemHelper({ element: "content" })}>{children}</span> : null}
      <div className={bemHelper({ element: "buttons" })}>
        {isDismissHidden ? null : (
          <Button
            kind={Button.kinds.TERTIARY}
            color={Button.colors.ON_PRIMARY_COLOR}
            className={cx(bemHelper({ element: "dismiss" }), dismissClassName)}
            size={Button.sizes.SMALL}
            {...otherDismissButtonProps}
          >
            {dismissContent || DISMISS_BUTTON_TEXT}
          </Button>
        )}
        {isSubmitHidden ? null : (
          <Button
            kind={Button.kinds.PRIMARY}
            color={Button.colors.ON_PRIMARY_COLOR}
            size={Button.sizes.SMALL}
            className={cx(bemHelper({ element: "submit" }), submitClassName)}
            {...otherSubmitButtonProps}
          >
            {submitContent || SUBMIT_BUTTON_TEXT}
          </Button>
        )}
      </div>
    </TipseenBasicContent>
  );
};

TipseenContent.propTypes = {
  title: PropTypes.string,
  onSubmit: PropTypes.func,
  onDismiss: PropTypes.func,
  isDismissHidden: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  isSubmitHidden: PropTypes.bool,
  // eslint-disable-next-line react/forbid-prop-types
  dismissButtonProps: PropTypes.object,
  // eslint-disable-next-line react/forbid-prop-types
  submitButtonProps: PropTypes.object
};

TipseenContent.defaultProps = {
  title: undefined,
  children: null,
  onSubmit: NOOP,
  onDismiss: NOOP,
  isDismissHidden: true,
  isSubmitHidden: false,
  dismissButtonProps: {},
  submitButtonProps: {}
};

export default TipseenContent;
