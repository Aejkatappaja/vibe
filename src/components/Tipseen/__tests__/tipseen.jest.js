import React from "react";
import { fireEvent, render, cleanup } from "@testing-library/react";
import { act } from "@testing-library/react-hooks";
import Tipseen from "../tipseen";
import renderer from "react-test-renderer";
import { TIPSEEN_CLOSE_BUTTON_TEST_ID } from "../TipseenConstants";

jest.useFakeTimers();
const renderComponent = ({ ...props }) => {
  return render(
    <Tipseen {...props}>
      <div />
    </Tipseen>
  );
};
describe("Tipseen tests", () => {
  describe("Snapshot Tests", () => {
    it("renders correctly with empty props", () => {
      const tree = renderer
        .create(
          <Tipseen>
            <div />
          </Tipseen>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders correctly with title", () => {
      const tree = renderer.create(<Tipseen title="mock title" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("renders correctly without close button", () => {
      const tree = renderer
        .create(
          <Tipseen isCloseButtonHidden>
            <div />
          </Tipseen>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("renders correctly without title", () => {
      const tree = renderer.create(<Tipseen />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("Integration Tests", () => {
    afterEach(() => {
      cleanup();
    });

    it("call onClose function when click on close button", () => {
      const onClickMock = jest.fn();
      const tipseen = renderComponent({
        onClose: onClickMock
      });
      const closeButton = tipseen.getByTestId(TIPSEEN_CLOSE_BUTTON_TEST_ID);

      act(() => {
        fireEvent.click(closeButton);
      });
      expect(onClickMock.mock.calls.length).toBe(1);
    });
  });
});
