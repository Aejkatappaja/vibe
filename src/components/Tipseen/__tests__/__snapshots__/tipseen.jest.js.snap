// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tipseen tests Snapshot Tests renders correctly without title 1`] = `
<div
  id=""
/>
`;
